repaired the range bug and added shared into Makefile

sneedcat opened the pull request:
Thanks to @hydrastro for helping me out.

sneedcat committed:
squashed changes

sneedcat committed:
added shared

skullchap reviewed (changes requested):
segfaults with ranges like (0, -20) in a string like "0123456789abcdef" where length < absolute len of start and end range.

Commit:
logic around negative ranges both in start and end

try this as a test suite

```c
// #define HUMAN_RANGE
#include "chadstr.h"

int main(void)
{
    puts(str(*str(range(str("0123456789abcdef"), 0,5))));
    puts(str(*str(range(str("0123456789abcdef"), 1,5))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), 0,0))));
    puts(str(*str(range(str("0123456789abcdef"), 5,0))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), 0,-3))));
    puts(str(*str(range(str("0123456789abcdef"), 13,-3))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), 0,-15))));
    puts(str(*str(range(str("0123456789abcdef"), 1,-15))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), 0,-16))));
    puts(str(*str(range(str("0123456789abcdef"), 1,-16))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), 0,-20))));
    puts(str(*str(range(str("0123456789abcdef"), 1,-20))));
    puts("");

    puts(str(*str(range(str("0123456789abcdef"), -1,-13))));
    puts(str(*str(range(str("0123456789abcdef"), -7,-13))));
    puts("");
}
```
